server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES93BloomFilterStoredFieldsFormat.java

```diff
@@ -0,0 +1,539 @@
+ *
+ * <ol>
+ *   <li>
+ *       <p>Bloom filter data file (extension .sfbf). This file stores the bloom
+ *       filter bitset.
+ *   <li>
+ *       <p>A bloom filter meta file (extension .sfbfm). This file stores metadata about the
+ *       bloom filters stored in the bloom filter data file. The in-memory representation can
+ *       be found in {@link BloomFilterMetadata}.
+ * </ol>
+ */
+public class ES93BloomFilterStoredFieldsFormat extends StoredFieldsFormat {
+    public static final String STORED_FIELDS_BLOOM_FILTER_FORMAT_NAME = "ES93BloomFilterStoredFieldsFormat";
+    public static final String STORED_FIELDS_BLOOM_FILTER_EXTENSION = "sfbf";
+    public static final String STORED_FIELDS_METADATA_BLOOM_FILTER_EXTENSION = "sfbfm";
+    private static final int VERSION_START = 1;
+    private static final int VERSION_CURRENT = VERSION_START;
+
+    // We use prime numbers with the Kirsch-Mitzenmacher technique to obtain multiple hashes from two hash functions
+    private static final int[] PRIMES = new int[] { 2, 5, 11, 17, 23, 29, 41, 47, 53, 59, 71 };
+    private static final int[] powerOfTwoBitSetSizes;
+    private static final int DEFAULT_NUM_HASH_FUNCTIONS = 7;
+    private static final byte BLOOM_FILTER_STORED = 1;
+    private static final byte BLOOM_FILTER_NOT_STORED = 0;
+
+    static {
+        // Precompute powers of two (2^1 to 2^26) for efficient modulo operations using bitwise AND.
+        // We start from 2^1 (2 bits) and go up to 2^26 (67,108,864 bits / 8,388,608 bytes = 8 MB)
+        // as the maximum, staying within positive int range.
+        powerOfTwoBitSetSizes = new int[27];
+        for (int i = 0; i < powerOfTwoBitSetSizes.length; i++) {
+            powerOfTwoBitSetSizes[i] = 1 << i;
+            assert powerOfTwoBitSetSizes[i] > 0;
+        }
+    }
+
+    private final BigArrays bigArrays;
+    private final String segmentSuffix;
+    private final StoredFieldsFormat delegate;
+    private final String bloomFilterFieldName;
+    private final int numHashFunctions;
+    private final int bloomFilterSizeInBits;
+
+    public ES93BloomFilterStoredFieldsFormat(
+        BigArrays bigArrays,
+        String segmentSuffix,
+        StoredFieldsFormat delegate,
+        ByteSizeValue bloomFilterSize,
+        String bloomFilterFieldName
+    ) {
+        this.bigArrays = bigArrays;
+        this.segmentSuffix = segmentSuffix;
+        this.delegate = delegate;
+        this.bloomFilterFieldName = bloomFilterFieldName;
+        this.numHashFunctions = DEFAULT_NUM_HASH_FUNCTIONS;
+        int bloomFilterSizeInBits = 0;
+        // Find the closest power of 2 that fits the required size
+        for (int powerOfTwoBitSetSize : powerOfTwoBitSetSizes) {
+            if (powerOfTwoBitSetSize <= (Math.multiplyExact(bloomFilterSize.getBytes(), Byte.SIZE))) {
+                bloomFilterSizeInBits = powerOfTwoBitSetSize;
+            }
+        }
+        assert bloomFilterSizeInBits > 0;
```

Comment:
```suggestion
        assert bloomFilterSizeInBits > 0 : bloomFilterSizeInBits;
```

Reply:
Tackled in https://github.com/elastic/elasticsearch/pull/137331/commits/a1af7b966a245fdceacabf1e723402564fc6ec1a